### DISCUSSION TITLE
How much audio data one needs for VITS2 to start generalizing well?

### DISCUSSION THREAD

**Discussion by michal-bida** [p0p4k/vits2_pytorch#96] [General]
Hi all.

I am using VITS2 to train a Czech voice. My question is - from your experience - how much audio data one needs to feed into the model the get a decent performance on out-of-domain words (not seen in training). For sure that will be language dependent - but any experience welcome. 

As for my experiments... So I have this low quality dataset of 40 minutes of Czech voice - out of that you get a decently sounding Czech voice after a 12 hours of training on A100. However, this voice absolutely does not generalize well - but performs well on in domain words.

Another dataset I have is 4 hours of a bit better quality of Czech voice - generalizes a bit better. However, it is by no means perfect. I am suspecting I will need perhaps twice the amount of data - maybe more - for the voice to generalize well.

Also, it seems to me that when you train the model with audio of different lengths then the model in inference kind of prefers the "knowledge" with duration that matches better the required duration... Concrete example:

- In your train set you have a longer sentence in gollum voice: "Gollum likes his precious ring very much so."
- But you have a lot of short sentences like so: "I liked the food." So word liked is represented well and you would expect the model to learn it properly. 

Then in inference you want to synthetize this sentence: "I liked his photo that he showed me yesterday."
You would expect "liked" to be pronounced correctly as you had plenty of examples. 
But what happened to me is that the synthetized voice sounded like: "I likes his photo that he showed me yesterday." - even though there were plenty of audio examples with word liked in the dataset. 

I think it is because 1) "liked his" somehow aligned better in the model to "likes his" and 2) the length of synthetized sentence were aligned better with the length of the "Gollum likes his precious ring very much so.". 

This leads me to thinking that perhaps if one wants a good generalization, one should use sentences of similar lengths in the training data? Alternatively, this might simply go away when more data are present...

EDIT: I should mention that this occurred after 4715 iterations - I will try to train more to see if it persists. It might go away.

What are your experiences with VITS2 performance based on the amount and structure of training data?

**Comment by p0p4k** [comment]
Are you using phonetic transcription? See if "liked his" and "likes his" end up giving same phones, if so, need to correct it manually.

**Reply by michal-bida** [reply]
Thanks for quick reply. Yes, I do use IPA transcription and it is actually different for both of these examples! It differs in one character ( to show the real Czech example, the IPA actually is `ktˈeriː nˈaːs` vs `ktˈer̝iː nˈaːs` - not only it is one character but it is a variant of r... ). (The English example is made up to make more sense for non Czech speaker.)
I am curious if it will actually go away with more iterations. Currently I am at 4700 iterations from 20K. 

EDIT Using ~5 hours of audio data.

**Reply by p0p4k** [reply]
Just train on the wrong sounding samples for few iterations.

**Reply by michal-bida** [reply]
So I have a follow up on this. This concerns IPA characters that are represented by several bytes. 
The whole issue with the above was that the model did not distinguish between `r̝̊` and `r` character. See:

```
phonemized_text = "pr̝̊ˈeji vˈaːm krˈaːsneː rˈaːno."

text_norm = text_to_sequence(phonemized_text, hps.data.text_cleaners)
text_norm = torch.LongTensor(text_norm)

print(text_norm)
tensor([ 58,  60, 156,  47,  52,  51,  16,  64, 156,  43, 158,  55,  16,  53,
         60, 156,  43, 158,  61,  56,  47, 158,  16,  60, 156,  43, 158,  56,
         57,   4])
phonemized_text = "prˈeji vˈaːm krˈaːsneː rˈaːno."

text_norm = text_to_sequence(phonemized_text, hps.data.text_cleaners)
text_norm = torch.LongTensor(text_norm)

print(text_norm)
tensor([ 58,  60, 156,  47,  52,  51,  16,  64, 156,  43, 158,  55,  16,  53,
         60, 156,  43, 158,  61,  56,  47, 158,  16,  60, 156,  43, 158,  56,
         57,   4])
```

Now you would think that the solution is to add r̝̊ character to the symbols right? Not so fast... This character consists actually of three characters and so far I haven't managed anything better than to have this character being represented by more symbols in the model. E.g.:

`r̝̊` = tensor([ 60, 179, 180]) instead of instead of tensor([ 60]).

I guess it would be better represented by just one symbol but this is a start...

@p0p4k do you know if there is a standardized way how to deal with these weird longer IPA characters such as `r̝̊` in Python?

TLDR:
When having issues with model unable to learn difference between characters check your symbol representation of the input!!

**Reply by michal-bida** [reply]
To answer myself. Yes, there is a way how to do it in Python by using regex library (as re does not deal well with these).

Something like

```
def split_into_graphemes(text):
    return regex.findall(r'\X', text)
```
will split String correctly including IPA characters. From here it shouldn't be a problem to rewrite symbols.py to support Czech IPA characters.

Thanks for providing this amazing implementation of VITS2 @p0p4k .

symbols.py for Czech:
```
""" from https://github.com/keithito/tacotron """
import regex

"""
Defines the set of symbols used in text input to the model.
"""

_pad = "_"
_punctuation = ';:,.!?¡¿—…"«»“” '
_letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
_letters_ipa = "ɑɐɒæɓʙβɔɕçɗɖðʤəɘɚɛɜɝɞɟʄɡɠɢʛɦɧħɥʜɨɪʝɭɬɫɮʟɱɯɰŋɳɲɴøɵɸθœɶʘɹɺɾɻʀʁɽʂʃʈʧʉʊʋⱱʌɣɤʍχʎʏʑʐʒʔʡʕʢǀǁǂǃˈˌːˑʼʴʰʱʲʷˠˤ˞↓↑→↗↘'̩'ᵻ"
_letters_ipa_cz = "r̝̊l̩r̩r̝"


unique_graphemes = set()
graphemes = regex.findall(r'\X', _pad + _punctuation + _letters + _letters_ipa + _letters_ipa_cz)
unique_graphemes.update(graphemes)
symbols = sorted(unique_graphemes)

# Special symbol ids
SPACE_ID = symbols.index(" ")
```

These two methods needs to be change in text.init:

```
import regex

def split_into_graphemes(text):
    return regex.findall(r'\X', text)


def text_to_sequence(text, cleaner_names):
    """Converts a string of text to a sequence of IDs corresponding to the symbols in the text.
    Args:
      text: string to convert to a sequence
      cleaner_names: names of the cleaner functions to run the text through
    Returns:
      List of integers corresponding to the symbols in the text
    """
    sequence = []

    clean_text = _clean_text(text, cleaner_names)
    graphemes = split_into_graphemes(clean_text)
    for symbol in graphemes:
        if symbol in _symbol_to_id.keys():
            symbol_id = _symbol_to_id[symbol]
            sequence += [symbol_id]
        else:
            continue
    return sequence


def cleaned_text_to_sequence(cleaned_text):
    """Converts a string of text to a sequence of IDs corresponding to the symbols in the text.
    Args:
      text: string to convert to a sequence
    Returns:
      List of integers corresponding to the symbols in the text
    """
    sequence = []

    graphemes = split_into_graphemes(cleaned_text)
    for symbol in graphemes:
        if symbol in _symbol_to_id.keys():
            symbol_id = _symbol_to_id[symbol]
            sequence += [symbol_id]
        else:
            continue
    return sequence
```

**Reply by p0p4k** [reply]
@michal-bida does Czech work on your trained model? How is the quality?

**Reply by michal-bida** [reply]
Czech absolutely works. Quality is good. Haven't yet trained for full 20K epoch yet but the results after thousands of epoch on several hours of curated Czech audio data were reasonably good (minus the r/ř character confusion caused by the issue above).

**Reply by p0p4k** [reply]
I wonder how cfm based methods compare? Cause now it is clear that your dataset is high quality.

**Reply by michal-bida** [reply]
I would be interested as well, though if your aim is the highest quality of generated voice then I would think using CFM would not help.

**Reply by p0p4k** [reply]
Why not? I thought cfm is the sota for non auto regressive spectrogram generation?

**Reply by michal-bida** [reply]
~Do you know about any recent models  using CFM? I wasn't able to find anything so far.~

**Reply by michal-bida** [reply]
Ok. For example https://arxiv.org/html/2309.03199v2

I read the paper.(Thanks for mentioning CFM btw). It seems to me that it indeed holds promise but from the paper I understood that the main advantage of matcha CFM model versus VITS2 would be probably smaller model size and speed and not so much the quality (MOS scores were on par with VITS and VITS2 should produce better results). 

Ofc speed of inference and training are important as well and I am pretty sure that the quality of speech could be also further improved.

Edit2: I see you already stated to implement this model https://github.com/p0p4k/Matcha-TTS-2 :-D

**Reply by p0p4k** [reply]
Or rather use pflow tts or matcha original, matcha-2 doesn't work yet, it is a wip.

**Reply by michal-bida** [reply]
Just FYI - if that would be interesting. So I tried the original matcha ( https://github.com/shivammehta25/Matcha-TTS ) on ~10 hours dataset and 3 ~hours czech datasets and I have to say that VITS2 is better when it comes to quality. Not sure if it can be caused by the used multilingual hifigan (I used some pretrained default version from Hugging face).

**Reply by p0p4k** [reply]
I think that continuous flow matching's simplicity and popularity must mean that it is better than the discrete normalizing flow. But something about vits' e2e training makes it superior i think...
